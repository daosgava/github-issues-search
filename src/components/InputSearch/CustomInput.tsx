import React, { ReactElement } from 'react';
import { Input } from '../shared';

type CustomInputProps = {
  handleOnChange: (e: React.FormEvent<HTMLInputElement>) => void;
  inputValue: string;
  handleKeyDown: (e: React.KeyboardEvent) => void;
  placeholder: string;
};

const CustomInput = ({
  inputValue,
  handleOnChange,
  handleKeyDown,
  placeholder,
}: CustomInputProps): ReactElement => {
  return (
    <>
      <Input
        tabIndex={0}
        type="text"
        value={inputValue}
        onChange={handleOnChange}
        onKeyDown={handleKeyDown}
        placeholder={placeholder}
      />
    </>
  );
};

export default CustomInput;
