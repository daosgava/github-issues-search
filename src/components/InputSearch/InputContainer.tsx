import React, { ReactElement } from 'react';
import { Input } from '../shared';

type InputContainerProps = {
  handleOnChange: (e: React.FormEvent<HTMLInputElement>) => void;
  inputValue: string;
  handleKeyDown: (e: React.KeyboardEvent) => void;
};

const InputContainer = ({
  inputValue,
  handleOnChange,
  handleKeyDown,
}: InputContainerProps): ReactElement => {
  return (
    <>
      <Input
        tabIndex={0}
        type="text"
        value={inputValue}
        onChange={handleOnChange}
        onKeyDown={handleKeyDown}
      />
    </>
  );
};

export default InputContainer;
