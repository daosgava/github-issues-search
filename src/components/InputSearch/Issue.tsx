import React, { ReactElement, useCallback } from 'react';
import Styled from 'styled-components';
import { node } from '../shared/types';

type styleProps = {
  ['data-highlight']: boolean;
}

const Li = Styled.li`
  cursor: pointer;
  margin: 2px;
  overflow: hidden;
  padding: 2px;
  background-color: ${(props: styleProps) => (props['data-highlight'] ? 'dodgerblue' : 'none')};
  color: ${(props: styleProps) => (props['data-highlight'] ? 'white' : 'black')};

  &:hover, &:focus{
    background-color: dodgerblue;
    border-radius: 2px;
    color: white;
    outline: 0;
  }
`;

const EmojiContainer = Styled.div`
  display: inline-block;
  margin-right: 10px;
`;

type IssueProps = {
  data: node;
  highlight: boolean;
};

const Issue = ({ data, highlight }: IssueProps): ReactElement => {

  const handleClick = useCallback(() => {
    window.location.assign(data.node.url);
  }, [data]);

  return (
    <Li
      onClick={handleClick}
      data-highlight={highlight}
    >
      <EmojiContainer>
        <span role="img" aria-label="search-result">🕵🏻‍♂️ </span>
      </EmojiContainer>
      {data.node.title}
    </Li>
  )
};

export default Issue;
