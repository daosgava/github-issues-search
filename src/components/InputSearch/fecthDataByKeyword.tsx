
import { node } from '../shared/types';
import axios from 'axios';
import { getGithubToken } from '../shared/utils';

const bearerToken = `bearer ${getGithubToken()}`;
// TODO: move routes to a config file
const endPoint = 'https://api.github.com/graphql';

const setQuery = (query: string, numberOfResults: number): string => (
  JSON.stringify({
    query:
      `query FindIssues {
      search(query: "${query}", type: ISSUE, first: ${numberOfResults}){
        issueCount
        edges{
          node{
            ... on Issue{
              id
              title
              url
            }
          }
        }
      }
    }`
  })
);

const fecthDataByKeyword = async (keyword: string, numberOfResults: number): Promise<node[]> => {
  if (!bearerToken) {
    throw new Error('Token not defined!');
  }

  if (!keyword) {
    return [];
  }

  try {
    const { data }: any = await axios.post(
      endPoint,
      setQuery(`repo:facebook/react is:issue ${keyword}`, numberOfResults),
      {
        headers: {
          Authorization: bearerToken,
        },
      },
    );

    return data.data.search.edges;
  } catch (e) {
    /*
      Log the error coming from the API
      but we don't break the page
    */
    console.log(e);

    return [];
  }
};

export default fecthDataByKeyword;
