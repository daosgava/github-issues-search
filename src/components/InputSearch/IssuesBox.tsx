import React, { ReactElement } from 'react';
import { node } from '../shared/types';
import { Box } from '../shared';
import Issue from './Issue';

type IssuesBoxProps = {
  issues: node[];
  cursor: number;
};

const IssuesBox = ({
  issues,
  cursor,
}: IssuesBoxProps): ReactElement => {
  return (
    <Box data-display={issues.length}>
      {
        issues.length > 0
        && issues.map((issue, index) => (
          <Issue key={`issue-${index}`} data={issue} highlight={cursor === index} />
        ))
      }
    </Box>
  );
};

export default IssuesBox;
