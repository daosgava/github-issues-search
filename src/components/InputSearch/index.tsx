import React, { ReactElement, useCallback, useState } from 'react';
import { node } from '../shared/types';
import CustomInput from './CustomInput';
import IssuesBox from './IssuesBox';
import { useDebounce, useFecthDataByKeyword } from '../shared/hooks';

const InputSearch = (): ReactElement => {
  const [inputValue, setInputValue] = useState<string>('');
  const [cursor, setCursor] = useState<number>(0);
  const numberOfResults = 10;
  const debouncedSearchTerm = useDebounce(inputValue, 500);
  const issues: node[] = useFecthDataByKeyword(debouncedSearchTerm, numberOfResults);

  const handleOnChange = useCallback((e: React.FormEvent<HTMLInputElement>): void => {
    setInputValue(e.currentTarget.value);
  }, []);

  const handleKeyDown = useCallback((e: React.KeyboardEvent): void => {
    if (e.key === 'Tab') {
      e.preventDefault();
    }

    if (e.key === 'ArrowUp' && cursor > 0) {
      setCursor(cursor - 1);
    } else if (e.key === 'ArrowDown' && cursor < numberOfResults - 1) {
      setCursor(cursor + 1);
    } else if (e.key === 'Enter') {
      window.location.assign(issues[cursor].node.url);
    }
  }, [cursor, issues]);

  return (
    <>
      <CustomInput
        inputValue={inputValue}
        handleOnChange={handleOnChange}
        handleKeyDown={handleKeyDown}
      />
      <IssuesBox issues={issues} cursor={cursor} />
    </>
  );
};

export default InputSearch;
