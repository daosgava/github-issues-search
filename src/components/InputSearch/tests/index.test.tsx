import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import InputSearch from '../index';

const setup = () => {
  const utils = render(<InputSearch />);
  const input = utils.getByPlaceholderText('Search for an issue') as HTMLInputElement;
  return {
    input,
    ...utils,
  }
};

test('InputSearch should match the snapshot', async () => {
  const { container } = setup();
  expect(container).toMatchSnapshot();

  await act(() => Promise.resolve());
});

test('Input value should change', async () => {
  const { input } = setup();
  fireEvent.change(input, { target: { value: 'bug' } });
  expect(input.value).toBe('bug');

  fireEvent.change(input, { target: { value: 'async' } });
  expect(input.value).toBe('async');

  await act(() => Promise.resolve());
});

