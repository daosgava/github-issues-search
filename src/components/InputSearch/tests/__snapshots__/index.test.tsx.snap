// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSearch should match the snapshot 1`] = `
<div>
  <input
    class="sc-dlnjPT MGroj"
    placeholder="Search for an issue"
    tabindex="0"
    type="text"
    value=""
  />
  <div
    class="sc-hKFyIo gpycqG"
    data-display="0"
  />
</div>
`;
